File: forwarder/pairing_test.go
package main

import (
	"strings"
	"testing"

	"github.com/kudzutechnologies/analytics/client"
)

func TestRenderPairConfigIncludesCredentialsAndExtras(t *testing.T) {
	config := defaultConf
	config.Endpoint = "analytics.example.com:50051"
	config.PairingEndpoint = "https://console.example.com/"
	config.CAFile = "/tmp/ca.pem"

	ini := renderPairConfig(config, &client.PairingConfig{
		GatewayID: "aabbccddeeff0011",
		ClientID:  "1122334455667788",
		ClientKey: "deadbeef",
		Extras: map[string]interface{}{
			"log-level":          "debug",
			"client-id":          "attacker-client",
			"analytics-endpoint": "attacker.example.com:443",
			"pairing-endpoint":   "https://attacker.example.com",
			"analytics-ca-file":  "/tmp/attacker.pem",
		},
	})

	for _, want := range []string{
		"client-id=1122334455667788",
		"client-key=deadbeef",
		"gateway=aabbccddeeff0011",
		"analytics-endpoint=analytics.example.com:50051",
		"pairing-endpoint=https://console.example.com/",
		"analytics-ca-file=/tmp/ca.pem",
		"log-level=debug",
	} {
		if !strings.Contains(ini, want) {
			t.Fatalf("rendered INI missing %q:\n%s", want, ini)
		}
	}
	for _, blocked := range []string{"attacker-client", "attacker.example.com", "attacker.pem"} {
		if strings.Contains(ini, blocked) {
			t.Fatalf("rendered INI contains protected extra %q:\n%s", blocked, ini)
		}
	}
}

func TestDefaultEndpoints(t *testing.T) {
	if defaultConf.Endpoint != "ingress.eu1.cluster.kudzu.gr:443" {
		t.Fatalf("analytics endpoint = %q", defaultConf.Endpoint)
	}
	if defaultConf.PairingEndpoint != "https://console.eu1.cluster.kudzu.gr" {
		t.Fatalf("pairing endpoint = %q", defaultConf.PairingEndpoint)
	}
}

File: forwarder-cc/tests/pairing_test.cc
#include "config.h"
#include "pairing.h"
#include "tests/test_harness.h"

using namespace forwarder_cc;
using namespace forwarder_cc_test;

int main() {
  Config endpoints;
  endpoints.endpoint = "analytics.example.com:50051";
  endpoints.pairing_endpoint = "https://console.example.com/";
  client_cc::PairingConfig paired;
  paired.gateway_id = "gateway-id";
  paired.client_id = "client-id";
  paired.client_key = "client-key";
  paired.extras["client-id"] = "attacker-client";
  paired.extras["analytics-endpoint"] = "attacker.example.com:443";
  paired.extras["pairing-endpoint"] = "https://attacker.example.com";
  std::string rendered = RenderPairConfig(endpoints, paired);
  Expect(rendered.find("pairing-endpoint=https://console.example.com/\n") !=
             std::string::npos,
         "rendered config preserves custom pairing endpoint");
  Expect(rendered.find("attacker") == std::string::npos,
         "protected pairing extras are ignored");

  // Pairing with invalid PIN should fail without network for empty digits...
  Config c;
  c.endpoint = "ingress.eu1.cluster.kudzu.gr:443";
  std::string err;
  std::string ini = GetRenderedPairConfig("abc", c, err);
  Expect(ini.empty(), "non-digit pin fails");
  Expect(!err.empty(), "error message set");

  // Render path is covered indirectly; ensure defaults leave gateway-* empty in DefaultConfig
  // by checking that a successful pair isn't required for this smoke test.
  ExpectEq(std::string(kForwarderVersion), "0.1.12", "version constant");

  return Summary();
}
